google-api-client/src/main/java/com/google/api/client/googleapis/services/EndToEndChecksumHandler.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.api.client.googleapis.services;
+
+import com.google.common.base.Ascii;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.logging.Logger;
+import javax.xml.bind.DatatypeConverter;
+
+/** This class provides End-to-End Checksum API for http protocol.
+ */
+public class EndToEndChecksumHandler {
+  private static final Logger log =
+      Logger.getLogger(EndToEndChecksumHandler.class.getCanonicalName());
+
+  /** The checksum http header on http requests */
+  public static final String HTTP_REQUEST_CHECKSUM_HEADER = "x-request-checksum-348659783";
+  /** The checksum http header on http responses */
+  public static final String HTTP_RESPONSE_CHECKSUM_HEADER = "x-response-checksum-348659783";
+
+  /**
+   * Create and return checksum as a string value for the input 'bytes'.
```

Comment:
why is a checksum a string?

Reply:
good point. made it byte[] and it removed some unnecessary stuff also

Reply:
good point. made it byte[]. seems to be way more accurate than making it string.